### DISCUSSION TITLE
[Help]Method overrides class in an incompatible manner (stuck for days and really need a hand)

### DISCUSSION THREAD

**Discussion by raceychan** [microsoft/pyright#6352] [Q&A] [answered]
# Intro
Dear community, I have been coding in python for years but just rencently started to take the typing system of python seriously, and as soon as i changed pyright to strict mode, I identified this certain type of problems existing everywhere in my application which pyright would compain:

Method  overrides class in an incompatible manner

## Problem description
I have an abstract parent class that defines all different kind of functions subclasses should have, but sometimes subclass can ask for a more specific type of the input. The rationale is that the parent class might take a dataclass as the argument, but subclasses might need a subclass of that dataclass with extra information to be able to complete the action.I hope the following code would help illustrate the issue i've been facing

## Example Code

```python 
import typing as ty


class Command:
    entity_id: str


class UserCommand(Command):
    user_name: str


class RemoveUser(UserCommand):
    ...


class CreateUser(UserCommand):
    user_password_hash: str


ActorCommands = ty.TypeVar("ActorCommands", bound=Command)


class Actor(ty.Generic[ActorCommands]):
    def create_child(self, command: ActorCommands) -> None:
        ...


UserCommands = ty.TypeVar("UserCommands", bound=UserCommand)


class UserActor(Actor[UserCommands]):
    duty: set[UserCommands]

    def create_child(self, command: CreateUser) -> None:
        validate(command.user_password_hash) # Here type checker would complain if I use UserCommands instead of CreateUser here.
```

Would anyone kind enough to give me a hand? I have been trying to use generic to solve this problem and got stuck for days, what is worse that at this point, I am still not sure if I am solving a problem that should not exist, does the problem exist because of the way I code and there is a better way to achieve this without generic type?

**Comment by erictraut** [comment]
Your sample is close to correct. The only problem is that in the method `UserActor.create_child`, the parameter `command` should be annotated using the type variable `UserCommands` rather than `CreateUser`.

**Reply by raceychan** [reply]
Thanks for your qucik response, eric.  As what you just suggested, I did try to use UserCommands, but it does not quite fit in my situation, for example, every `UserCommand` is supposed to contain user_id, but only `CreateUser` might contain extra information like user_password_hash.  I feel like Generic is the right path to take, since I've been looking at other libraries try to find a solution and found many other libraries use generic, but I always feel like I am missing something.

**Reply by raceychan** [reply]
I just updated my example and hope that would better illustrate my dilemma here.
Another thing is that, as I was looking at python official docs for `typing.TypeVar`,  at [typing](https://docs.python.org/3/library/typing.html),

in the section of class typing.TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False)

the docs indicates that when use bound=type in TypeVar, subtype of type should also work:
```python
T = TypeVar('T')  # Can be anything
S = TypeVar('S', bound=str)  # Can be any subtype of str
A = TypeVar('A', str, bytes)  # Must be exactly str or bytes
```

and that is why I am creating a `UserCommand` and bind it to TypeVar `UserCommands`, hope that `CreateUser` as a subtype of `UserCommand` would fit in the `UserActor.create_child` signature, but somehow the `Method overrides class in an incompatible manner` would still show up.

**Reply by erictraut** [reply, marked as answer]
It looks like you're trying to define a class hierarchy that is intrinsically unsafe from a type perspective. If you want a solution that is type safe, you may need to revisit your design. I don't know enough about the problem you're trying to solve, so I can't offer a specific solution.

Perhaps this will help clarify why your code sample above is not type safe. Your implementation of `UserActor.create_child` assumes that the argument passed to the `command` parameter is a `CreateUser` instance. Let's say that you have an instance of `UserActor[UserCommand]`. This is a subtype of `Actor[UserCommand]`, which means it should be possible to use a `UserActor[UserCommand]` anywhere you can use an `Actor[UserCommand]`. However, the `Actor[UserCommand].create_child` method accepts a `command` parameter that is of type `UserCommand`. If you pass a `UserCommand` as an argument, the implementation of `create_child` in the `UserActor` class will crash when it attempts to access the `user_password_hash` attribute.

**Reply by raceychan** [reply]
Thanks Eric, I get your point, I konw that subclass should override parent class methods only with more general argument and more specific return type to comply with the liskov substitution principle, and my current design is breaking it, but I just thought generic can somehow fix this without redesigning the code. 
Your response makes me realize that I do need to reconsider the design,  thank you, i do appreciate that!

**Reply by raceychan** [reply]
> I don't know enough about the problem you're trying to solve, so I can't offer a specific solution.

Eric, I just came up with a simpler example that might help demonstrate the problem I am trying to solve, would you be able to take took at this?

```python
import typing as ty


Convertable = ty.TypeVar('Convertable', int, float, str)


class Converter(ty.Generic[Convertable]):
    def convert(self, value: Convertable) -> bool:
        raise NotImplementedError


class StringConverter(Converter[str]):
    def convert(self, value: str) -> bool:
        return value != "false"

class IntConverter(Converter[int]):
    def convert(self, value: int) -> bool:
        return value != 0
```

I think the concept is similar to template partial specialization in cpp, but with a little extra complexity where there is inheritance involve in Convertable, so we might have a `SubConvertable` where constraints of `SubConvertable` are subtypes of contrains in `Convertable`
Is this possible to achieve if I put it this way? or It just a whole different situation when inheritance got involved.

**Reply by erictraut** [reply]
This example is quite different from your previous example. This code is type safe (and therefore no type errors are generated). If this approach addresses the problem you're trying to solve, then that's great.

I'll note that your use of a constrained type variable here looks a bit strange. You're specifying both `int` and `float` as constraints, but `int` is a subtype of `float`, so `float` already implies `int`. You could remove the `int`, and it would work the same.
